java/ql/src/experimental/Security/CWE/CWE-522/UnsecureBasicAuth.ql

```diff
@@ -0,0 +1,228 @@
+ */
+class URLConstructorTaintStep extends TaintTracking::AdditionalTaintStep {
+  override predicate step(DataFlow::Node node1, DataFlow::Node node2) {
+    exists(URLConstructor u |
+      node1.asExpr() = u.stringArg() and
+      node2.asExpr() = u
+    )
+  }
+}
+
+/**
+ * Tracks the flow of data from `openConnection` method to the connection object.
+ */
+class OpenHttpURLTaintStep extends TaintTracking::AdditionalTaintStep {
+  override predicate step(DataFlow::Node node1, DataFlow::Node node2) {
+    exists(MethodAccess ma, VariableAssign va |
+      ma.getMethod() instanceof HttpURLOpenMethod and
+      ma.getQualifier() = node1.asExpr() and
+      (
+        ma = va.getSource()
+        or
+        exists(CastExpr ce |
+          ce.getExpr() = ma and
+          ce = va.getSource() and
+          ce.getControlFlowNode().getASuccessor() = va //With a type cast like HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+        )
+      ) and
+      node2.asExpr() = va.getDestVar().getAnAccess()
+    )
+  }
+}
+
+class HttpStringToHttpURLOpenMethodFlowConfig extends TaintTracking::Configuration {
+  HttpStringToHttpURLOpenMethodFlowConfig() {
+    this = "UnsecureBasicAuth::HttpStringToHttpURLOpenMethodFlowConfig"
+  }
+
+  override predicate isSource(DataFlow::Node src) { src.asExpr() instanceof HttpString }
+
+  override predicate isSink(DataFlow::Node sink) {
+    sink.asExpr().(VarAccess).getVariable().getType() instanceof TypeUrlConnection or
+    sink.asExpr().(VarAccess).getVariable().getType() instanceof TypeHttpUrlConnection // TypeHttpUrlConnection isn't instance of TypeUrlConnection
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
+    exists(DataFlow::Node nodem |
+      any(URLConstructorTaintStep uts).step(node1, nodem) and
+      any(OpenHttpURLTaintStep ots).step(nodem, node2)
+    )
+  }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    node.getType() instanceof PrimitiveType or node.getType() instanceof BoxedType
+  }
+}
+
+/**
+ * The method `setRequestProperty()` declared in URL Connection invoked for basic authentication.
+ */
+class SetBasicAuthPropertyMethodAccess extends MethodAccess {
+  SetBasicAuthPropertyMethodAccess() {
+    this.getMethod().getDeclaringType() instanceof TypeUrlConnection and
+    this.getMethod().hasName("setRequestProperty") and
+    this.getArgument(0).(CompileTimeConstantExpr).getStringValue() = "Authorization" and
+    builtFromBasicAuthStringConcat(this.getArgument(1)) and
+    exists(VarAccess conn, DataFlow::PathNode source, DataFlow::PathNode sink, HttpString s |
+      this.getQualifier() = conn and //HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+      source.getNode().asExpr() = s and
+      sink.getNode().asExpr() = conn.getVariable().getAnAccess() and
+      any(HttpStringToHttpURLOpenMethodFlowConfig c).hasFlowPath(source, sink)
+    )
```

Comment:
I think there is no reason to restrict the qualifier to be a `VarAccess`. 
```suggestion
    exists(DataFlow::PathNode source, DataFlow::PathNode sink, HttpString s |
      this.getQualifier() = conn and //HttpURLConnection conn = (HttpURLConnection) url.openConnection();
      source.getNode().asExpr() = s and
      sink.getNode().asExpr() = this.getQualifier() and
      any(HttpStringToHttpURLOpenMethodFlowConfig c).hasFlowPath(source, sink)
    )
```

Why did you add the comment `//HttpURLConnection conn = (HttpURLConnection) url.openConnection();` instead of stating `this.getQualifier() instance of `TypeHttpUrlConnection` ?

Reply:
The method has been replaced by `BasicAuthFlowConfig` of the type `TaintTracking::Configuration`, which is not restricted to `VarAccess`.